entities/sync/sync.go

```diff
@@ -9,104 +9,173 @@
-func NewKeyLocker() *KeyLocker {
-	return &KeyLocker{
-		m: sync.Map{},
-	}
+// KeyLocker provides thread-safe locking by key with automatic resource cleanup.
+// It maintains a bounded number of locks and automatically removes unused ones
+// based on both size limits and time-based expiration.
+//
+// Features:
+//   - Thread-safe key-based locking
+//   - Automatic cleanup of unused locks
+//   - Size-based eviction (LRU)
+//   - Time-based expiration
+//   - Memory leak prevention
+type KeyLocker struct {
+	locks   sync.Map
+	maxAge  time.Duration // maximum age of unused locks
+	count   atomic.Int32  // current count of locks
+	maxSize int32         // maximum number of locks
+	logger  logrus.FieldLogger
 }
 
-// Lock it locks a specific bucket by it's ID
-// to hold ant concurrent access to that specific item
+// NewKeyLocker creates a new KeyLocker with default settings.
+// It initializes a KeyLocker with:
+//   - Maximum size of 1000 locks
+//   - Maximum age of 24 hours for unused locks
+//   - Automatic cleanup every 5 minutes
 //
-//	do not forget calling Unlock() after locking it.
-func (s *KeyLocker) Lock(ID string) {
-	iLock := &sync.Mutex{}
-	iLocks, _ := s.m.LoadOrStore(ID, iLock)
+// The created KeyLocker automatically starts a background goroutine
+// for periodic cleanup of expired locks.
+func NewKeyLocker(logger logrus.FieldLogger) *KeyLocker {
+	kl := &KeyLocker{
+		maxSize: 1000,           // default max size
+		maxAge:  24 * time.Hour, // default max age
+		logger:  logger,
+	}
 
-	iLock = iLocks.(*sync.Mutex)
-	iLock.Lock()
-}
+	enterrors.GoWrapper(func() { // cleanup every 5 minutes
+		kl.periodicCleanup(5 * time.Minute)
+	}, kl.logger)
 
-// Unlock it unlocks a specific item by it's ID
-func (s *KeyLocker) Unlock(ID string) {
-	iLocks, _ := s.m.Load(ID)
-	iLock := iLocks.(*sync.Mutex)
-	iLock.Unlock()
+	return kl
 }
 
-// KeyRWLocker it is a thread safe wrapper of sync.Map
-// Usage: it's used in order to lock/rlock specific key in a map
-// to synchronize concurrent access to a code block.
-//
-//	locker.Lock(id)
-//	defer locker.Unlock(id)
+// Lock acquires a lock for the given key. If the key doesn't exist,
+// a new lock is created. If the maximum size is reached, the oldest
+// unused lock is evicted before creating a new one.
 //
-// or
+// The lock is guaranteed to be held when this method returns. The caller
+// must ensure they call Unlock when they're done with the lock.
 //
-//	locker.RLock(id)
-//	defer locker.RUnlock(id)
-type KeyRWLocker struct {
-	m sync.Map
-}
+// If the key already exists, its last-used timestamp is updated,
+// preventing it from being cleaned up while in use.
+func (kl *KeyLocker) Lock(key string) {
+	now := time.Now().UnixNano()
 
-// NewKeyLocker creates Keylocker
-func NewKeyRWLocker() *KeyRWLocker {
-	return &KeyRWLocker{
-		m: sync.Map{},
+	// Check if we already have this lock
+	if value, ok := kl.locks.Load(key); ok {
+		entry := value.(*lockEntry)
```

Comment:
there is a theoretical race at this line where entry is returned from the map, however, we haven't updated the `lastUsed`. The lock can be removed from the map and then cause undefined behaviour when releasing

Reply:
good point